### PR TITLE
Create BE Profile Controllers(#6) and Model for Profile(#7)

### PR DESCRIPTION
**Pull request by lesurafel**
Create Profile Model.

### CODE REVIEW

**Comment by jerixmx** [comment]
Hello, Surafel! Congratulations on submitting your first code PR for this repo!

I have some comments to improve your pull request:
- Use a self-explanatory Pull Request Title.
- Use the Pull Request Template as this helps with documentation and review of changes. This template should be available when you create your pull request on GitHub.
- Connect/link issues to your pull request. Ideally, your pull request should have a single responsibility so that issues can move along independently.
- Use a branch name that describes the scope of your request. [Resource](https://codingsight.com/git-branching-naming-convention-best-practices/)
- Use more meaningful commit messages. [Resource](https://dev.to/i5han3/git-commit-message-convention-that-you-can-follow-1709)

Please update the title of this pull request and fill out the pull request template to improve it. Looking forward to your next PRs!

File: server/models/Profile.js
Comment (jerixmx) [inline server/models/Profile.js:8]: Please use camelCase which is also consistent with the `userSchema`

File: server/models/Profile.js
Comment (jerixmx) [inline server/models/Profile.js:5]: Isn't the validator `required`, not `require`?

File: server/models/Profile.js
Comment (jerixmx) [inline server/models/Profile.js:16]: The email is already in the User model. You can populate the `userId` instead when you need the email to avoid conflicts.

File: server/models/Profile.js
Comment (jerixmx) [inline server/models/Profile.js:21]: Since `gender` is a dropdown in the mocks, you can use `enum` to specify the allowed values for gender

File: server/models/Profile.js
Comment (jerixmx) [inline server/models/Profile.js:24]: :+1:

File: server/models/Profile.js
Comment (jerixmx) [inline server/models/Profile.js:23]: I'm curious - any  reason the phone number is a string array instead of just string?

File: server/models/Profile.js
Comment (jerixmx) [inline server/models/Profile.js:38]: If this array is for storing both the profile photo and the other images that appear in the bio, I suggest that you store the profilePhoto here as a String and have another field for the bio images that stores the URL to those images as a String array. 

You may also want to set a default profile photo.

File: server/models/Profile.js
Comment (jerixmx) [inline server/models/Profile.js:37]: Can you expound on your scheme for storing availability? Are these days of the week and number of hours per day? The hours may have to be split to start and end times.

File: server/routes/profile.js
Comment (jerixmx) [inline server/routes/profile.js:13]: No need to specify `create` for this POST request.

File: server/routes/profile.js
Comment (jerixmx) [inline server/routes/profile.js:16]: Instead of adding `update` to path, specify the ID of the profile to be updated for this PUT request.

File: server/routes/profile.js
Comment (jerixmx) [inline server/routes/profile.js:19]: Please document testing of these routes. You can do this through Postman.

File: server/app.js
Comment (jerixmx) [inline server/app.js:14]: You can actually omit `.js` :)

File: server/controllers/profile.js
Comment (jerixmx) [inline server/controllers/profile.js:9]: You can get the `userId` from `req.user.id` then check validity with the User database. Getting this from the request body risks profile creation by another user.

File: server/controllers/profile.js
Comment (jerixmx) [inline server/controllers/profile.js:22]: You can check first whether the user already has an existing profile to avoid duplication or conflict.

File: server/controllers/profile.js
Comment (jerixmx) [inline server/controllers/profile.js:43]: Good job specifying the creation status!

File: server/controllers/profile.js
Comment (jerixmx) [inline server/controllers/profile.js:93]: I suggest to use `profile` variable name instead of `profiles` for clarity because this should return just one profile.

File: server/models/Profile.js
Comment (jerixmx) [inline server/models/Profile.js:3]: Nit: Spelling this out as `requiredString` would be better than `reqString` as `reqString` seems to mean `request string`.

File: server/controllers/profile.js
Comment (jerixmx) [inline server/controllers/profile.js:67]: :+1:

File: server/controllers/profile.js
Comment (jerixmx) [inline server/controllers/profile.js:88]: The user should only be able to update their own profile. You can use `req.user.id` instead of getting it from the query.

File: server/controllers/profile.js
Comment (jerixmx) [inline server/controllers/profile.js:91]: It would be best to verify first that the query fields match those in the Profile schema instead of directly using the query to update the profile.

**Review by jerixmx — CHANGES_REQUESTED** [review]
Hi Surafel, I've made some comments on your code. You can sweep through most of these! Feel free to respond to these comments for clarifications. 

Please include screenshots of API responses on each route (you can do this through Postman) when you fill out the pull request template.

Happy coding!

**Comment by lesurafel** [comment]
> Hello, Surafel! Congratulations on submitting your first code PR for this repo!
> 
> I have some comments to improve your pull request:
> 
> * Use a self-explanatory Pull Request Title.
> * Use the Pull Request Template as this helps with documentation and review of changes. This template should be available when you create your pull request on GitHub.
> * Connect/link issues to your pull request. Ideally, your pull request should have a single responsibility so that issues can move along independently.
> * Use a branch name that describes the scope of your request. [Resource](https://codingsight.com/git-branching-naming-convention-best-practices/)
> * Use more meaningful commit messages. [Resource](https://dev.to/i5han3/git-commit-message-convention-that-you-can-follow-1709)
> 
> Please update the title of this pull request and fill out the pull request template to improve it. Looking forward to your next PRs!

Hello Jeri, Thank you for your comment! I will fix all the comments.

File: server/models/Profile.js
Comment (lesurafel) [reply server/models/Profile.js:23]: May be the user has more than one phone number. In the document that I got from you in dropbox, the Edit Profile picture shows the system allows the user can put more than one phone number. I do based on that.

File: server/models/Profile.js
Comment (lesurafel) [reply server/models/Profile.js:37]: No these days are not the day of the week. The days variable store the specific dates that selected by the user from the calendar.

File: server/models/Profile.js
Comment (jerixmx) [reply server/models/Profile.js]: I see! With that intent, your data type makes sense. :) 

Here's my copy of the Edit Profile mock. Usually, we only have one number associated to an account. Here, it appears that there's only an `Add a phone number button` since there's no associated phone number yet.
setup-profile-edit
`

File: server/models/Profile.js
Comment (jerixmx) [reply server/models/Profile.js]: Great! You can use `dates` instead of `days` and [Date] instead of [String]. You can improve this field by making your availability scheme more self-explanatory for others who would be using this model.

Comment (moffatethan) [reply server/models/Profile.js]: We will talk more about this @lesurafel on our meeting tomorrow on the best approach here for a better UI/UX. 😄

**Comment by moffatethan** [comment]
@lesurafel 

"Please include screenshots of API responses on each route (you can do this through Postman) when you fill out the pull request template." This is not needed, starting today requirements for back-end screenshots are removed and only in front-end work should you include screenshots. Thank you!

File: server/controllers/profile.js
Comment (moffatethan) [inline server/controllers/profile.js:37]: Would this ever be hit? We should be doing middleware validation on the incoming request data to ensure reliability and keep our controller logic lean.

File: server/controllers/profile.js
Comment (moffatethan) [reply server/controllers/profile.js:41]: Good on using the right status, I don't see the need to pass in a success boolean and a data payload property, I would just return the profile data `res.status(201).json({ profile })` 

It's simpler, we can validate on the front-end whether a request was successful indicated by the 2xx HTTP status codes.

File: server/controllers/profile.js
Comment (moffatethan) [inline server/controllers/profile.js:48]:
```suggestion
exports.findAllProfiles = asyncHandler(async (req, res, next) => {
```

Please be mindful or proper whitespacing 😄

File: server/controllers/profile.js
Comment (moffatethan) [inline server/controllers/profile.js:54]: This is unnecessary, profiles should always be an empty array, this is actually a successful response because we've fetched all profiles and there are no profiles indicated by an empty array of there is indicated by a populated array.

File: server/controllers/profile.js
Comment (moffatethan) [inline server/controllers/profile.js:59]: Same as before, this is redundant and we could just be passing in the profiles

File: server/controllers/profile.js
Comment (moffatethan) [inline server/controllers/profile.js:78]: Same here as above, we could just pass in the profile

File: server/controllers/profile.js
Comment (moffatethan) [inline server/controllers/profile.js:88]: Good job here

File: server/controllers/profile.js
Comment (moffatethan) [inline server/controllers/profile.js:91]: We're doing two database requests here

File: server/controllers/profile.js
Comment (moffatethan) [reply server/controllers/profile.js:88]: To address my comment below, you can add in `{ new: true }` argument into the findOneAndUpdate and it will return with the updated document which means we don't have to do two database requests

File: server/models/Profile.js
Comment (moffatethan) [inline server/models/Profile.js:51]: We will discuss this in our meeting tomorrow

**Review by moffatethan — APPROVED** [review]
Left you some comments @lesurafel, looks decent just some touch ups need to be done before approval.

**Review by moffatethan — APPROVED** [review]
@lesurafel Good to go! 

**Please address merge conflicts then merge**

**Pull request merged**
